File: EMS/EMS/EMSEventsParser.m
//
//  EMSEventsParser.m
//

#import "EMSEventsParser.h"
#import "EMSConference.h"

#import "CJCollection.h"
#import "CJLink.h"
#import "CJItem.h"

#import "EMSDateConverter.h"

@implementation EMSEventsParser

- (NSArray *)processData:(NSData *)data andHref:(NSURL *)href error:(NSError **)error {
    NSError *parseError = nil;

    CJCollection *collection = [CJCollection collectionForNSData:data error:&parseError];

    if (!collection) {
        DDLogError(@"Failed to retrieve conferences %@ - %@ - %@", href, parseError, [parseError userInfo]);

        *error = parseError;

        return [NSArray array];
    }

    NSMutableArray *temp = [[NSMutableArray alloc] init];

    [collection.items enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        CJItem *item = (CJItem *) obj;

        EMSConference *conf = [[EMSConference alloc] init];

        conf.href = item.href;

        [item.data enumerateObjectsUsingBlock:^(id dataObj, NSUInteger dataIdx, BOOL *dataStop) {
            NSDictionary *dict = (NSDictionary *) dataObj;

            NSString *field = dict[@"name"];
            NSString *value = dict[@"value"];

            if ([@"name" isEqualToString:field]) {
                conf.name = value;
            }
            if ([@"venue" isEqualToString:field]) {
                conf.venue = value;
            }
            if ([@"start" isEqualToString:field]) {
                conf.start = [EMSDateConverter dateFromString:value];
            }
            if ([@"end" isEqualToString:field]) {
                conf.end = [EMSDateConverter dateFromString:value];
            }
        }];

        [item.links enumerateObjectsUsingBlock:^(id linksObj, NSUInteger linksIdx, BOOL *linksStop) {
            CJLink *link = (CJLink *) linksObj;

            if ([@"session collection" isEqualToString:link.rel]) {
                conf.sessionCollection = link.href;

                if (link.otherFields != nil) {
                    conf.hintCount = link.otherFields[@"count"];
                }
            }
            if ([@"slot collection" isEqualToString:link.rel]) {
                conf.slotCollection = link.href;
            }
            if ([@"room collection" isEqualToString:link.rel]) {
                conf.roomCollection = link.href;
            }
        }];

        [temp addObject:conf];
    }];

    return [NSArray arrayWithArray:temp];
}

- (void)parseData:(NSData *)data forHref:(NSURL *)href {
    NSError *error = nil;

    NSArray *collection = [self processData:data andHref:href error:&error];

    [self.delegate finishedEvents:collection forHref:href error:error];
}

@end

File: EMS/Network/Parsers/EMSEventsParser.swift
import Foundation
import CocoaLumberjack

@objc protocol EMSEventsParserDelegate {
    optional func finishedEvents(conferences: [EMSConference], forHref href:NSURL, error:NSError?)
}

@objc class EMSEventsParser : NSObject {
    var delegate : EMSEventsParserDelegate?
    
    func processData(data: NSData, forHref href:NSURL) throws -> [EMSConference] {
        
        do {
            let collection = try CJCollection(forNSData:data)
            
            return collection.items.map({ item in
                let conference = EMSConference()
                
                conference.href = item.href

                for datum in item.data {
                    if let field = datum["name"] as? String, value = datum["value"] as? String {
                        switch (field) {
                        case "name":
                            conference.name = value
                        case "venue":
                            conference.venue = value
                        case "start":
                            conference.start = EMSDateConverter.dateFromString(value)
                        case "end":
                            conference.end = EMSDateConverter.dateFromString(value)
                        default:
                            DDLogInfo("Unknown field \(field) for conference")
                        }
                    }
                }
                
                for link in item.links {
                    switch (link.rel) {
                        case "session collection":
                        conference.sessionCollection = link.href
                        case "slot collection":
                        conference.slotCollection = link.href
                        case "room collection":
                        conference.roomCollection = link.href
                    default:
                        DDLogInfo("Unknwon rel \(link.rel) for conference")
                    }
                    
                    // Ugly - otherFields is coming in as implicitly unwrapped optional of dictionary
                    if let count = link.otherFields!["count"] as? NSNumber {
                        conference.hintCount = count
                    }
                }

                return conference
            })
        } catch let error as NSError {
            DDLogError("Failed to retrieve conferences \(href) - \(error)")
            
            throw error
        }
    }
    
    func parseData(data: NSData, forHref url: NSURL) {
        do {
            let collection = try self.processData(data, forHref: url)
            
            self.delegate?.finishedEvents?(collection, forHref: url, error: nil)
        } catch let error as NSError {
            self.delegate?.finishedEvents?([], forHref: url, error: error)
        }
    }
}
